### ISSUE TITLE
Define an abstract list table base class, used to extend all list tables.

### ISSUE THREAD

**Issue by ramiabraham** [AffiliateWP/AffiliateWP#1528]
Defining an abstract list table base class allows us to unify this object plugin-wide, and share various methods and features while keeping things _dry_.

**Comment by ramiabraham** [comment]
Thoughts on this? Kept it as simple as possible for now, though there might be _somewhat_ defined handlers for optional ajax methods, export functionality, and search filters later on

**Comment by pippinsplugins** [comment]
`get_views()` and `_get_views()` are not quite right.

`get_views()` should be used by each sub class and should do nothing more than setup the array of views with key values pairs.
 
`_get_view()` should call `get_views()` and then handle the logic for setting up the HTML, class names, filters, etc.

The same goes for the other methods, get_columns, _get_columns(), get_sortable_columns(), and _get_sortable_columns().

**Comment by ramiabraham** [comment]
Something similar to what I had in the prior iteration (last commit that had this: https://github.com/AffiliateWP/AffiliateWP/commit/c2642f192ff3ec2cf6d050f8137d3ee5411e9a1f)?

**Comment by pippinsplugins** [comment]
Not quite. Like this:

Base class:

```php
/**
 * Retrieve the table columns.
 *
 * @access public
 * @since  1.9
 * @return array $columns Array of all the list table columns.
 * @uses   $this->_get_columns()
 */
public function get_columns() {
	$columns = array();
	return $columns;
}

/**
 * Set the table columns.
 *
 * @access public
 * @since  1.9
 * @return array $columns Array of all the list table columns.
 */
private function _get_columns() {

	$columns = $this->get_columns();

	/**
	 * This filter defines the table columns for this list table.
	 *
	 * The filter name is structured as:
	 *    1. Prefix: affwp_list_table.
	 *    2. The singular form of the object name, eg 'affiliate'.
	 *    3. Suffix: _table_columns
	 *
	 *    Example:
	 *    affwp_list_table_affiliate_table_columns
	 *
	 * @since  1.9
	 */
	return apply_filters( 'affwp_list_table_' . $this->singular . '_table_columns', $columns );
}
```

In sub class:

```php
/**
 * Retrieve the table columns.
 *
 * @access public
 * @since  1.9
 * @return array $columns Array of all the list table columns.
 * @uses   $this->_get_columns()
 */
public function get_columns() {
	$columns = array(
		'id' => __( 'ID' ),
		'name' => __( 'Name' )
	);
	return $columns;
}
```

**Comment by DrewAPicture** [comment]
This was implemented and completed as part of #1525.

Columns can be hidden via a `columns_to_hide` argument passed to the `display_args` argument key in the list table constructor, directly:

```php
	// Recent Payouts.
	$payouts_table = new AffWP_Payouts_Table( array(
		'query_args' => array(
			'affiliate_id' => $affiliate_id
		),
		'display_args' => array(
			'hide_bulk_options'  => true,
			'columns_to_hide' => array( 'status' ),
		),
	) );
	$payouts_table->prepare_items();
	?>
	<h2><?php _e( 'Recent Payouts', 'affiliate-wp' ); ?></h2>

	<?php $payouts_table->views(); ?>
	<?php $payouts_table->display(); ?>
```
